cloneDeviceRecord shares empty (non-nil) Metadata map between original and clone

# Summary
- **Context**: `cloneDeviceRecord` in `pkg/registry/device_store.go` is used throughout the device registry to create defensive copies of device records, ensuring that callers can safely modify returned records without affecting the original stored data.
- **Bug**: When cloning a `DeviceRecord` with an empty (but not nil) `Metadata` map, the function fails to deep-copy the map, causing both the original and clone to share the same underlying map reference.
- **Actual vs. expected**: Modifications to the cloned record's `Metadata` map directly affect the original stored record's `Metadata`, instead of the two records having independent metadata maps.
- **Impact**: Data corruption in the device registry's in-memory store, where concurrent or sequential operations on "independent" device record copies can unexpectedly overwrite each other's metadata changes, leading to incorrect device state and potential race conditions.

# Code with bug

```go
func cloneDeviceRecord(src *DeviceRecord) *DeviceRecord {
	if src == nil {
		return nil
	}

	dst := *src

	if src.Hostname != nil {
		hostname := *src.Hostname
		dst.Hostname = &hostname
	}
	if src.MAC != nil {
		mac := *src.MAC
		dst.MAC = &mac
	}
	if src.IntegrationID != nil {
		integrationID := *src.IntegrationID
		dst.IntegrationID = &integrationID
	}
	if src.CollectorAgentID != nil {
		collectorAgentID := *src.CollectorAgentID
		dst.CollectorAgentID = &collectorAgentID
	}

	if len(src.DiscoverySources) > 0 {
		dst.DiscoverySources = append([]string(nil), src.DiscoverySources...)
	}
	if len(src.Capabilities) > 0 {
		dst.Capabilities = append([]string(nil), src.Capabilities...)
	}
	if len(src.Metadata) > 0 {  // <-- BUG 🔴 This condition is false for empty maps
		meta := make(map[string]string, len(src.Metadata))
		for k, v := range src.Metadata {
			meta[k] = v
		}
		dst.Metadata = meta
	}

	return &dst
}
```

The bug is at line 204 in `pkg/registry/device_store.go`. When `src.Metadata` is an empty map (length 0 but not nil), the condition `if len(src.Metadata) > 0` evaluates to false, so the map is not cloned. Since `dst := *src` creates a shallow copy of the struct, `dst.Metadata` still points to the same underlying map as `src.Metadata`.

# Evidence

## Example

Consider the following sequence:

1. A `DeviceRecord` is stored in the registry with `Metadata: map[string]string{}`
2. Caller A retrieves the record via `GetDeviceRecord()`, which returns a clone
3. Caller A modifies the clone: `clone.Metadata["key1"] = "value1"`
4. Caller B retrieves the same record via `GetDeviceRecord()`, which returns another clone
5. **BUG**: Caller B's clone already contains `key1 = "value1"` even though they expect a fresh copy

This happens because:
- Step 1: The registry stores record R with empty map M at address 0x1000
- Step 2: `cloneDeviceRecord(R)` returns clone C1, but C1.Metadata still points to 0x1000 (not copied)
- Step 3: Caller A writes to 0x1000["key1"]
- Step 4: `cloneDeviceRecord(R)` returns clone C2, but C2.Metadata still points to 0x1000
- Step 5: Both R.Metadata, C1.Metadata, and C2.Metadata point to the same map at 0x1000

## Failing test

### Test script

```go
package registry

import (
	"testing"
)

// TestCloneDeviceRecord_EmptyMapIsolation verifies that cloning a DeviceRecord
// with an empty (but not nil) Metadata map creates an independent copy.
func TestCloneDeviceRecord_EmptyMapIsolation(t *testing.T) {
	// Create a record with an empty (but not nil) metadata map
	original := &DeviceRecord{
		DeviceID: "device-1",
		IP:       "10.0.0.1",
		Metadata: map[string]string{}, // Empty map, not nil
	}

	// Clone the record
	clone := cloneDeviceRecord(original)

	// Modify the clone's metadata
	clone.Metadata["test_key"] = "test_value"

	// The original should NOT be affected
	if len(original.Metadata) > 0 {
		t.Errorf("BUG: Modifying clone's metadata affected the original! Original.Metadata = %v", original.Metadata)
	}

	// The clone should have the new value
	if clone.Metadata["test_key"] != "test_value" {
		t.Errorf("Expected clone to have test_key=test_value, got: %v", clone.Metadata)
	}
}

// TestGetDeviceRecord_ModifyingCloneAffectsOriginal_EmptyMap demonstrates
// a realistic scenario where the cloning bug causes data corruption.
func TestGetDeviceRecord_ModifyingCloneAffectsOriginal_EmptyMap(t *testing.T) {
	reg := newTestDeviceRegistry()

	// Insert a device with an empty (but not nil) metadata map
	originalRecord := &DeviceRecord{
		DeviceID: "device-1",
		IP:       "10.0.0.1",
		Metadata: map[string]string{}, // Empty but not nil
	}
	reg.UpsertDeviceRecord(originalRecord)

	// Retrieve the device (which should return a clone)
	clone1, ok := reg.GetDeviceRecord("device-1")
	if !ok {
		t.Fatal("Expected to find device-1")
	}

	// Verify the clone starts empty
	if len(clone1.Metadata) != 0 {
		t.Fatalf("Expected clone to have empty metadata, got: %v", clone1.Metadata)
	}

	// Modify the clone's metadata
	clone1.Metadata["key1"] = "value1"

	// Retrieve the device again
	clone2, ok := reg.GetDeviceRecord("device-1")
	if !ok {
		t.Fatal("Expected to find device-1")
	}

	// BUG: The second clone should NOT have the modifications from clone1
	if val, exists := clone2.Metadata["key1"]; exists {
		t.Errorf("BUG: Modifications to clone1 leaked into clone2! clone2.Metadata = %v", clone2.Metadata)
		t.Errorf("Expected clone2.Metadata to be empty, but found key1=%v", val)
	}
}

// TestGetDeviceRecord_RaceCondition shows how this bug could cause a race condition
func TestGetDeviceRecord_RaceCondition(t *testing.T) {
	reg := newTestDeviceRegistry()

	// Insert a device with empty metadata
	reg.UpsertDeviceRecord(&DeviceRecord{
		DeviceID: "device-1",
		IP:       "10.0.0.1",
		Metadata: map[string]string{},
	})

	// Two different callers retrieve the device
	caller1, ok1 := reg.GetDeviceRecord("device-1")
	caller2, ok2 := reg.GetDeviceRecord("device-1")

	if !ok1 || !ok2 {
		t.Fatal("Expected to find device-1")
	}

	// Both callers modify their "independent" copies
	caller1.Metadata["source"] = "caller1"
	caller2.Metadata["source"] = "caller2"

	// BUG: One caller's modification overwrites the other's
	if caller1.Metadata["source"] == "caller2" {
		t.Errorf("BUG: caller2's modification affected caller1's copy!")
	}
}
```

### Test output

Running `go test -v -count=1 ./pkg/registry -run TestCloneDeviceRecord_EmptyMapIsolation`:

```
=== RUN   TestCloneDeviceRecord_EmptyMapIsolation
    device_store_clone_test.go:25: BUG: Modifying clone's metadata affected the original! Original.Metadata = map[test_key:test_value]
--- FAIL: TestCloneDeviceRecord_EmptyMapIsolation (0.00s)
FAIL
FAIL	github.com/carverauto/serviceradar/pkg/registry	0.005s
FAIL
```

Running `go test -v -count=1 ./pkg/registry -run TestGetDeviceRecord_ModifyingCloneAffectsOriginal_EmptyMap`:

```
=== RUN   TestGetDeviceRecord_ModifyingCloneAffectsOriginal_EmptyMap
    device_store_realistic_bug_test.go:44: BUG: Modifications to clone1 leaked into clone2! clone2.Metadata = map[key1:value1]
    device_store_realistic_bug_test.go:45: Expected clone2.Metadata to be empty, but found key1=value1
--- FAIL: TestGetDeviceRecord_ModifyingCloneAffectsOriginal_EmptyMap (0.00s)
FAIL
FAIL	github.com/carverauto/serviceradar/pkg/registry	0.006s
FAIL
```

Running `go test -v -count=1 ./pkg/registry -run TestGetDeviceRecord_RaceCondition`:

```
=== RUN   TestGetDeviceRecord_RaceCondition
    device_store_realistic_bug_test.go:81: BUG: caller2's modification affected caller1's copy!
--- FAIL: TestGetDeviceRecord_RaceCondition (0.00s)
FAIL
FAIL	github.com/carverauto/serviceradar/pkg/registry	0.006s
FAIL
```

# Full context

The device registry maintains an in-memory cache of device records (`DeviceRegistry.devices map[string]*DeviceRecord`). To provide defensive copying and prevent external callers from corrupting this internal state, all getter methods (`GetDeviceRecord`, `FindDevicesByIP`, `FindDevicesByMAC`, `SnapshotRecords`) use `cloneDeviceRecord` to return deep copies of records.

Empty (but not nil) `Metadata` maps can be created in several ways:

1. **Direct initialization with empty maps**: Throughout the codebase, there are many instances where code initializes metadata as `map[string]string{}` (e.g., in tests at `pkg/registry/registry_test.go:284`, `pkg/registry/service_device_test.go:297`)

2. **Via `mergeMetadata` function**: In `pkg/registry/device_store_update.go:145-163`, the `mergeMetadata` function can produce empty maps when:
   - `dest` is nil
   - `src` has entries but they all have empty/whitespace-only keys
   - Example: `mergeMetadata(nil, map[string]string{"": "value"})` returns an empty (non-nil) map

3. **Via various update flows**: The registry processes device updates through multiple integration sources (Armis, Netbox, sweep discovery), and metadata can start as an initialized empty map (e.g., `pkg/registry/registry.go:1622`, `pkg/db/db.go:354`)

The bug affects any code path that:
1. Retrieves a device record using `GetDeviceRecord()`
2. Modifies the returned record's `Metadata` map
3. Expects those modifications to be isolated from other callers

Key affected call sites:
- `pkg/registry/device_store_update.go:34`: Gets existing record, passes to `deviceRecordFromUpdate` which may mutate it
- `pkg/registry/registry.go:1896`: Public API `GetDevice()` returns cloned records
- `pkg/registry/registry.go:1960`: Search functionality retrieves and operates on cloned records
- `pkg/registry/registry.go:2044`: `FindRelatedDevices()` retrieves cloned records

# Why has this bug gone undetected?

This bug has remained undetected for several reasons:

1. **Uncommon edge case**: Empty (but not nil) metadata maps are less common than either nil maps or maps with content. Most device records in production have at least some metadata entries, which would trigger the cloning logic and work correctly.

2. **Defensive pattern elsewhere**: The codebase has defensive checks like `if update.Metadata == nil { update.Metadata = make(map[string]string) }` (e.g., `pkg/registry/registry.go:1621-1623`) before writing to metadata. This pattern masks the bug because it creates a new map before writing, preventing corruption of the shared empty map.

3. **Existing tests don't validate isolation**: The existing test `TestUpsertAndGetDeviceRecord` at `pkg/registry/device_store_test.go:60-78` does test mutation isolation, but it uses a record with non-empty metadata (`metadata := map[string]string{"region": testRegionUSEast1}`), so the cloning logic works correctly in that test.

4. **Timing-dependent manifestation**: The bug only manifests when:
   - A record starts with empty (not nil) metadata
   - The record is retrieved multiple times
   - The cloned copies are modified
   - This specific sequence is less common in actual usage patterns

5. **Map reference equality not tested**: None of the existing tests verify that cloned records have independent map instances (by checking if modifications to one clone affect another), they only check value equality at a single point in time.

# Recommended fix

Change the condition at line 204 in `pkg/registry/device_store.go` from `if len(src.Metadata) > 0` to `if src.Metadata != nil`:

```go
if src.Metadata != nil {  // <-- FIX 🟢 Check for nil instead of length
	meta := make(map[string]string, len(src.Metadata))
	for k, v := range src.Metadata {
		meta[k] = v
	}
	dst.Metadata = meta
}
```

This ensures that any non-nil map (including empty maps) gets deep-copied, while truly nil maps remain nil in the clone.

The same issue likely affects the slice fields (`DiscoverySources` and `Capabilities`), though in practice this is less problematic because Go slice headers are copied by value, and the append operations used in the code typically create new underlying arrays. However, for consistency and safety, these should also be fixed:

```go
if src.DiscoverySources != nil {  // <-- FIX 🟢 Check for nil instead of length
	dst.DiscoverySources = append([]string(nil), src.DiscoverySources...)
}
if src.Capabilities != nil {  // <-- FIX 🟢 Check for nil instead of length
	dst.Capabilities = append([]string(nil), src.Capabilities...)
}
```

--- a/openspec/changes/fix-clone-device-record-metadata-aliasing/proposal.md
+++ b/openspec/changes/fix-clone-device-record-metadata-aliasing/proposal.md
@@ -1,0 +1,21 @@
+# Change: Fix `cloneDeviceRecord` metadata aliasing for empty maps
+
+## Why
+Issue #2145 reports that `pkg/registry/device_store.go:cloneDeviceRecord` fails to deep-copy `DeviceRecord.Metadata` when the source map is empty-but-non-nil. Because the record is shallow-copied (`dst := *src`), the clone and original end up sharing the same underlying map reference, defeating the defensive-copying contract used throughout the device registry.
+
+This can cause surprising cross-call contamination (a later clone “inherits” keys written to a previous clone), incorrect device state in the registry’s in-memory cache, and potential data races when callers concurrently read/write “independent” record copies.
+
+## What Changes
+- Update `cloneDeviceRecord` to deep-copy `Metadata` when `src.Metadata != nil` (including empty maps).
+- Update `cloneDeviceRecord` to deep-copy `DiscoverySources` and `Capabilities` when non-nil (including empty-but-non-nil slices) to avoid similar aliasing via shared backing arrays.
+- Add regression tests covering:
+  - Empty-but-non-nil `Metadata` map isolation (original vs clone and clone vs clone).
+  - Empty-but-non-nil slices with non-zero capacity (append to clone does not affect original or subsequent clones).
+
+## Impact
+- Affected specs: `device-registry-defensive-copying`
+- Affected code:
+  - `pkg/registry/device_store.go` (`cloneDeviceRecord`)
+  - `pkg/registry/*_test.go` (new/updated unit tests)
+- Risk: Low. Changes are limited to defensive copying; primary impact is small additional allocations during record cloning.
+

--- a/openspec/changes/fix-clone-device-record-metadata-aliasing/specs/device-registry-defensive-copying/spec.md
+++ b/openspec/changes/fix-clone-device-record-metadata-aliasing/specs/device-registry-defensive-copying/spec.md
@@ -1,0 +1,22 @@
+## ADDED Requirements
+
+### Requirement: DeviceRecord clone isolation
+The device registry MUST return `DeviceRecord` values whose mutable fields do not alias internal stored state or other returned clones, including when those fields are empty-but-non-nil.
+
+#### Scenario: Empty metadata maps are deep-copied
+- **GIVEN** a stored `DeviceRecord` has `Metadata` set to an empty (but non-nil) map
+- **WHEN** a caller retrieves a copy via a registry getter that uses `cloneDeviceRecord`
+- **THEN** the returned `Metadata` MUST be a distinct map instance from the stored record
+- **AND** mutating the returned `Metadata` MUST NOT change the stored record’s `Metadata`
+
+#### Scenario: Empty slices are non-aliased
+- **GIVEN** a stored `DeviceRecord` has `DiscoverySources` and/or `Capabilities` set to empty-but-non-nil slices (including cases with non-zero capacity)
+- **WHEN** a caller retrieves a copy via a registry getter that uses `cloneDeviceRecord`
+- **THEN** appending to the returned slices MUST NOT affect the stored record’s slices
+- **AND** subsequent clones MUST NOT observe values appended to earlier clones
+
+#### Scenario: Clone mutation does not affect other clones
+- **GIVEN** two callers retrieve independent clones of the same stored `DeviceRecord`
+- **WHEN** one caller mutates its clone’s `Metadata` (e.g., sets a new key)
+- **THEN** the other caller’s clone MUST NOT reflect that mutation
+

--- a/openspec/changes/fix-clone-device-record-metadata-aliasing/tasks.md
+++ b/openspec/changes/fix-clone-device-record-metadata-aliasing/tasks.md
@@ -1,0 +1,13 @@
+## 1. Fix cloning behavior
+- [x] 1.1 Update `cloneDeviceRecord` to deep-copy `Metadata` for any non-nil map (even when empty)
+- [x] 1.2 Update `cloneDeviceRecord` to deep-copy `DiscoverySources` and `Capabilities` for any non-nil slice (even when empty)
+- [x] 1.3 Ensure pointer fields (`Hostname`, `MAC`, `IntegrationID`, `CollectorAgentID`) remain non-aliased as today
+
+## 2. Add regression tests
+- [x] 2.1 Add unit test: empty-but-non-nil `Metadata` is deep-copied (mutating clone does not affect original)
+- [x] 2.2 Add unit test: two clones from the same record do not share `Metadata` (mutating one does not affect the other)
+- [x] 2.3 Add unit test: empty-but-non-nil slices (including `make([]string, 0, N)`) do not alias (appending to clone does not affect original or later clones)
+
+## 3. Verification
+- [x] 3.1 Run `go test ./pkg/registry/...`
+- [x] 3.2 Run `go test -race ./pkg/registry/...` (or a targeted race repro test if needed)

--- a/pkg/registry/device_store.go
+++ b/pkg/registry/device_store.go
@@ -195,13 +195,13 @@ func cloneDeviceRecord(src *DeviceRecord) *DeviceRecord {
 		dst.CollectorAgentID = &collectorAgentID
 	}
 
-	if len(src.DiscoverySources) > 0 {
-		dst.DiscoverySources = append([]string(nil), src.DiscoverySources...)
+	if src.DiscoverySources != nil {
+		dst.DiscoverySources = append(make([]string, 0, len(src.DiscoverySources)), src.DiscoverySources...)
 	}
-	if len(src.Capabilities) > 0 {
-		dst.Capabilities = append([]string(nil), src.Capabilities...)
+	if src.Capabilities != nil {
+		dst.Capabilities = append(make([]string, 0, len(src.Capabilities)), src.Capabilities...)
 	}
-	if len(src.Metadata) > 0 {
+	if src.Metadata != nil {
 		meta := make(map[string]string, len(src.Metadata))
 		for k, v := range src.Metadata {
 			meta[k] = v

--- a/pkg/registry/device_store_test.go
+++ b/pkg/registry/device_store_test.go
@@ -78,6 +78,63 @@ func TestUpsertAndGetDeviceRecord(t *testing.T) {
 	}
 }
 
+func TestCloneDeviceRecord_EmptyMetadataMapDoesNotAlias(t *testing.T) {
+	original := &DeviceRecord{
+		DeviceID: testDeviceID1,
+		IP:       "10.0.0.1",
+		Metadata: map[string]string{},
+	}
+
+	clone1 := cloneDeviceRecord(original)
+	if clone1 == nil {
+		t.Fatalf("expected clone to be non-nil")
+	}
+
+	clone1.Metadata["k"] = "v1"
+
+	if len(original.Metadata) != 0 {
+		t.Fatalf("expected original metadata to remain empty, got %#v", original.Metadata)
+	}
+
+	clone2 := cloneDeviceRecord(original)
+	if clone2 == nil {
+		t.Fatalf("expected second clone to be non-nil")
+	}
+	if _, ok := clone2.Metadata["k"]; ok {
+		t.Fatalf("expected second clone metadata to not include prior clone mutation, got %#v", clone2.Metadata)
+	}
+}
+
+func TestCloneDeviceRecord_EmptySlicesWithCapacityDoNotAlias(t *testing.T) {
+	original := &DeviceRecord{
+		DeviceID:         testDeviceID1,
+		IP:               "10.0.0.1",
+		DiscoverySources: make([]string, 0, 1),
+		Capabilities:     make([]string, 0, 1),
+	}
+
+	clone1 := cloneDeviceRecord(original)
+	if clone1 == nil {
+		t.Fatalf("expected clone to be non-nil")
+	}
+	clone1.DiscoverySources = append(clone1.DiscoverySources, "a")
+	clone1.Capabilities = append(clone1.Capabilities, "icmp")
+
+	clone2 := cloneDeviceRecord(original)
+	if clone2 == nil {
+		t.Fatalf("expected second clone to be non-nil")
+	}
+	clone2.DiscoverySources = append(clone2.DiscoverySources, "b")
+	clone2.Capabilities = append(clone2.Capabilities, "snmp")
+
+	if got := clone1.DiscoverySources[0]; got != "a" {
+		t.Fatalf("expected clone1 discovery source to remain %q, got %q", "a", got)
+	}
+	if got := clone1.Capabilities[0]; got != "icmp" {
+		t.Fatalf("expected clone1 capability to remain %q, got %q", "icmp", got)
+	}
+}
+
 func TestUpsertUpdatesIndexes(t *testing.T) {
 	reg := newTestDeviceRegistry()
 

